### PR TITLE
Aglez/dev/cache

### PR DESCRIPTION
**Pull request by arturogonzalez58**
Before this PR, the telepresence id was generated for each job using the action. This PR adds a cache layer to save the telepresence ID and re-use it in different jobs. 

This PR adds the configure action. The configure action must run in the main branch in order to propagate the telepresence id to the rest of PR/branch that uses the action.

The other actions pull the cache generated by the main branch to use the same telepresence ID, if the cache is not generated by the configuration action, the action fails.

This behavior was tested in the [development repository](github.com/datawire/development-telepresence-intercept) and the [test repository](https://github.com/arturogonzalez58/emojivoto/actions).

### CODE REVIEW

**Comment by hojenkai** [comment]
there is an `.idea` folder, we should probably remove it

File: configure/action.yaml [hunk 1]
```diff
@@ -0,0 +1,10 @@
+name: 'Configure Telepresence on the repository'
+description: 'Set the configuration to use telepresence in the repository'
+inputs:
+  version:
+    description: The version of telepresence to use
+    required: false
+    default: 2.5.8
```
Comment (hojenkai) [inline configure/action.yaml:7]: `latest` should probably be the default value right?

File: configure/configure.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    const restorePath = [path];
+
+    try {
+        await io.mkdirP(path);
+        await cache.restoreCache(restorePath, configure.TELEPRESENCE_CACHE_KEY);
+    } catch (error) {
+        core.warning(`Unable to find the telepresence id: ${error}`);
+    }
+    try {
+        await exec.exec(`${installTelepresence.TP_PATH}/telepresence`, ['connect']);
```
Comment (hojenkai) [inline configure/configure.js:21]: curious why we need to reference the `TP_PATH` here. Is it a custom path or the tp binary will not be available in the `path` anymore?

**Review by KeoniMuller — APPROVED** [review]

File: configure/configure.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+const io = require('@actions/io');
+const configure = require('../src/configure');
+const installTelepresence = require('../src/install');
+const cache = require('@actions/cache');
+
+const telepresenceConfiguring = async function () {
+    await installTelepresence.telepresenceInstall();
+
+    const path = configure.getTelepresenceConfigPath();
+    const restorePath = [path];
```
Comment (hojenkai) [inline configure/configure.js:12]:
```suggestion
    const telepresenceConfigDir = [path];
```

File: configure/configure.js (hunk 2 above)
Comment (hojenkai) [reply configure/configure.js:21]: looks like `core.addPath` should help us there to reference `telepresence` as we did

File: configure/configure.js [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    try {
+        await exec.exec(`${installTelepresence.TP_PATH}/telepresence`, ['connect']);
+    } catch (error) {
+        core.setFailed(error.message);
+        return;
+    }
+    try {
+        const cacheKey = await cache.saveCache(restorePath, configure.TELEPRESENCE_CACHE_KEY);
+        if (!cacheKey)
+            core.setFailed('Unable to save the cache');
```
Comment (hojenkai) [inline configure/configure.js:29]:
```suggestion
            core.setFailed('Unable to save the telepresence key cache');
```

File: src/configure/index.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+const core = require('@actions/core');
+const io = require('@actions/io');
+const cache = require('@actions/cache');
+
+exports.getTelepresenceConfigPath = () => {
+    switch (process.platform) {
+        case "darwin":
+            return `/home/runner/Library/Application\\ Support/telepresence`;
+        case "linux":
+            return `/home/runner/.config/telepresence`;
```
Comment (hojenkai) [inline src/configure/index.js:10]: no string interpolation done here

```suggestion
            return '/home/runner/.config/telepresence';
```

File: src/configure/index.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+const core = require('@actions/core');
+const io = require('@actions/io');
+const cache = require('@actions/cache');
+
+exports.getTelepresenceConfigPath = () => {
+    switch (process.platform) {
+        case "darwin":
+            return `/home/runner/Library/Application\\ Support/telepresence`;
```
Comment (hojenkai) [inline src/configure/index.js:8]: no string interpolation done here

```suggestion
            return '/home/runner/Library/Application\\ Support/telepresence';
```

File: src/configure/index.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+exports.getConfiguration = async () => {
+    const path = this.getTelepresenceConfigPath();
+    try {
+        await io.mkdirP(path);
+        const cacheid = await cache.restoreCache([path], this.TELEPRESENCE_CACHE_KEY,)
+        if (!cacheid)
+            core.setFailed('Unable to find a telepresence install id stored');
+    } catch (error) {
+        core.setFailed(error);
```
[the remark this answers is not in the document]
Comment (arturogonzalez58) [reply src/configure/index.js:25]: it should prevent the execution of other commands

File: configure/configure.js (hunk 2 above)
Comment (arturogonzalez58) [reply configure/configure.js:21]: I have some issues calling telepresence from there, that's why I'm using the path, let's try again this and we can decide if we keep it or remove it

File: src/install/index.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+        `https://app.getambassador.io/download/tel2/linux/amd64/${version}/telepresence`;
+
+    let tpCacheId = await cache.restoreCache([this.TP_PATH], cacheKey);
+
+    if (!tpCacheId) {
+        try {
+            await toolCache.downloadTool(TELEPRESENCE_DOWNLOAD_URL, `${this.TP_PATH}/telepresence`);
+            tpCacheId = await cache.saveCache([this.TP_PATH], cacheKey);
+            if(!tpCacheId)
+                core.setFailed('There was a problem saving the telepresence binary.');
```
Comment (hojenkai) [inline src/install/index.js:25]: `core.setFailed` doesn't stop the execution of the function. If we want to continue with the execution we'd probably want a `warning` instead

File: package.json [hunk 9]
```diff
@@ -19,8 +19,11 @@
   "homepage": "https://github.com/datawire/telepresence-actions#readme",
   "dependencies": {
     "@actions/artifact": "^1.1.0",
+    "@actions/cache": "^2.0.5",
     "@actions/core": "^1.8.2",
     "@actions/exec": "^1.1.1",
-    "@actions/http-client": "^2.0.1"
+    "@actions/http-client": "^2.0.1",
+    "@actions/io": "^1.1.2",
```
[the remark this answers is not in the document]
Comment (arturogonzalez58) [reply package.json:26]: I took a look at the package.json from core action and it doesn't include the io package. https://github.com/actions/toolkit/blob/8263c4d15d3a8616851e9c7762ce46fbb4f8c552/packages/core/package.json#L38-L42

**Pull request merged**